Support storing and applying of memory content

martin-fleck-at opened the pull request:
#### What it does
Add commands to store and apply memory content as Intel HEX file
- Encapsulate behavior in new MemoryStorage class
- Trigger 'store' from Memory view, Variables view and command palette
- Trigger 'apply' from Memory view, Explorer view and command palette
- Use nrf-intel-hex library for read/write file licensed under BSD-3

Use quick inputs to guide user through necessary input
- Initialize as much of the input as possible through command args

Communicate with webview through messenger requests and notifications
- Request to trigger store and apply from webview
- Notify webview about any written memory so it can update properly

Minor improvements
- Move some common types and functionality into 'common' area
- Avoid bleeding Debug Adapter types into webview, use messaging types
- Common style: 'getVariables' -> 'getVariablesType'
- Provide utility functions and types for debug requests
- Fix 'Enter' handling for numpad by checking key value of event

https://github.com/eclipse-cdt-cloud/vscode-memory-inspector/assets/19170971/9c3e2772-cfd5-4b7c-8d45-e2707338d0cb


Closes #50

#### How to test
- Run a debug session
- Check how you can store memory (from the inspector, from the command, from the Variables window)
- Try applying your stored memory (from the inspector, from the file context menu, from the file explorer context menu)

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

colin-grant-work commented on src/plugin/memory-webview-main.ts:
```diff
@@ -231,28 +233,40 @@ export class MemoryWebview implements vscode.CustomReadonlyEditorProvider {
+    protected async writeMemory(request: WriteMemoryArguments): Promise<WriteMemoryResult> {
         try {
-            return await this.memoryProvider.writeMemory(request);
+            return this.memoryProvider.writeMemory(request);
```
Inside a `try` block, the `await` is required; otherwise a rejection can't be caught here.

martin-fleck-at replied:
Great catch!

colin-grant-work replied:
No pun intended? :-)

martin-fleck-at replied:
Dang, I didn't even notice, I could have appeared so clever ;-)

colin-grant-work commented on src/webview/components/memory-table.tsx:
```diff
@@ -102,7 +103,7 @@ interface MemoryTableProps extends TableRenderOptions, MemoryDisplayConfiguratio
     count: number;
     effectiveAddressLength: number;
-    fetchMemory(partialOptions?: Partial<DebugProtocol.ReadMemoryArguments>): Promise<void>;
+    fetchMemory(partialOptions?: Partial<ReadMemoryArguments>): Promise<void>;
```
```suggestion
    fetchMemory(partialOptions?: MemoryOptions): Promise<void>;
```

colin-grant-work commented on src/webview/components/memory-widget.tsx:
```diff
@@ -32,14 +33,16 @@ interface MemoryWidgetProps extends MemoryDisplayConfiguration {
     updateTitle: (title: string) => void;
-    fetchMemory(partialOptions?: Partial<DebugProtocol.ReadMemoryArguments>): Promise<void>
+    fetchMemory(partialOptions?: MemoryOptions): Promise<void>;
+    triggerStoreMemory(): void;
+    triggerApplyMemory(): void;
```
All function calls trigger some action:

```suggestion
    storeMemory(): void;
    applyMemory(): void;
```

colin-grant-work commented on src/webview/memory-webview-view.tsx:
```diff
@@ -225,6 +256,17 @@ class App extends React.Component<{}, MemoryAppState> {
         this.setState(prevState => ({ ...prevState, isFrozen: !prevState.isFrozen }));
     }
 
+    protected requestStoreMemory = (): void => {
+        messenger.sendRequest(storeMemoryType, HOST_EXTENSION, {
+            memoryReference: this.state.memoryReference,
+            count: this.state.count,
+            offset: this.state.offset
+        });
+    };
+
+    protected requestApplyMemory = async (): Promise<void> => {
+        await messenger.sendRequest(applyMemoryType, HOST_EXTENSION, undefined);
+    };
```
We haven't used `request` in other cases where we make requests; that it's a request to the plugin side is a bit of an implementation detail.

```suggestion
    protected storeMemory = (): void => {
        messenger.sendRequest(storeMemoryType, HOST_EXTENSION, {
            memoryReference: this.state.memoryReference,
            count: this.state.count,
            offset: this.state.offset
        });
    };

    protected applyMemory = async (): Promise<void> => {
        await messenger.sendRequest(applyMemoryType, HOST_EXTENSION, undefined);
    };
```

martin-fleck-at commented on src/common/debug-requests.ts:
```diff
@@ -0,0 +1,39 @@
+import type { DebugProtocol } from '@vscode/debugprotocol';
+import type { DebugSession } from 'vscode';
+
+export interface DebugRequestTypes {
+    'evaluate': [DebugProtocol.EvaluateArguments, DebugProtocol.EvaluateResponse['body']]
+    'readMemory': [DebugProtocol.ReadMemoryArguments, DebugProtocol.ReadMemoryResponse['body']]
+    'writeMemory': [DebugProtocol.WriteMemoryArguments, DebugProtocol.WriteMemoryResponse['body']]
+}
```
Absolutely true! I wonder if the VS Code people had a discussion about that because as soon as you actually start using it, you really see how accessing the `body` of the responses makes for really long and really weird types ;-)

So my thought process in the split between the `debug-requests.ts` and `messaging.ts` is that the messaging is really the protocol between the extension's main side and the extension webview whereas the debug protocol messages are between the extension's main side and the debug adapter.

Therefore, bleeding any types from the debug protocol into the webview does not make much sense as it cannot directly access the debug adapter. Following that, it seemed logical to have a messaging layer in-between where we define the arguments and results of the main-webview communication. Sure, some of those types may be a 1:1 alias for a debug protocol message but we may also want to extend them at some point (like I do with the `WriteMemoryArguments`), make parts required or optional (with a lot of the `Partial` going around) or re-define them. So in a way, having all those `DebugProtocol.ReadMemoryArguments` in the webview and the options is really more a "coincidence" because it obviously represents memory reading well but and also avoids giving some of the usages a more specific semantic imho.

Based on that, I'm absolutely not opposed to using the `DebugProtocol` interfaces on the extension's main side within all the debug-related classes (which is why I also left them in the `memory-provider`). Of course I wished for their names to be shorter, so we could alias them as in the `debug-requests.ts` as you suggested but for me personally, that would not replace the type definitions that we do in `messaging.ts`.

Long story short: I think we should distinguish between the types used in the main-debug communication and the ones in the main-webview communication. However, maybe I just overthought the whole thing so I don't plan to die on this hill ;-) I'd be very happy to get your thoughts on that though.

martin-fleck-at commented on src/common/debug-requests.ts:
```diff
@@ -0,0 +1,39 @@
+    return session.customRequest(command, args);
+}
+
+export namespace EvaluateExpression {
+    export function sizeOf(expression: string): string {
+        return `sizeof(${expression})`;
+    }
+    export function addressOf(expression: string): string {
+        return `&(${expression})`;
+    }
+};
```
Ah, for some reason I thought that might be part of the protocol but I was wrong. Thank you for that, I'll move it to the adapter capabilities!

martin-fleck-at commented on src/plugin/memory-webview-main.ts:
```diff
@@ -15,41 +15,41 @@
+    StoreMemoryArguments,
+    WriteMemoryArguments,
+    WriteMemoryResult,
+    applyMemoryType,
+    getVariablesType,
     logMessageType,
-    setOptionsType,
+    memoryWrittenType,
     readMemoryType,
-    writeMemoryType,
-    MemoryReadResult,
-    MemoryWriteResult,
-    getVariables,
-    setMemoryViewSettingsType,
+    readyType,
     resetMemoryViewSettingsType,
+    setMemoryViewSettingsType,
+    setOptionsType,
     setTitleType,
+    storeMemoryType,
+    writeMemoryType,
```
I fully agree with you and thank you for going through this, especially with the number and size of the PRs! Some shared code styles would definitely elevate the pain here.

martin-fleck-at commented on src/plugin/memory-webview-main.ts:
```diff
@@ -190,15 +189,18 @@ export class MemoryWebview implements vscode.CustomReadonlyEditorProvider {
                     this.refresh(participant);
                 }
             }),
+            this.memoryProvider.onDidWriteMemory(writtenMemory => this.messenger.sendNotification(memoryWrittenType, participant, writtenMemory))
```
You are absolutely right! I was actually thinking about this as well but thought it might be too much for this PR. Currently we just notify all the memory inspectors and they'll reload their state if they match the changed range - even if they do not actually are about this session. Since this causes overhead due to more than necessary reads but not really any faults, thank you for creating a separate ticket for it!

martin-fleck-at commented on src/common/memory-range.ts:
```diff
@@ -93,8 +99,8 @@ export function getAddressLength(padding: number, radix: Radix): number {
 
-export function toHexStringWithRadixMarker(target: bigint): string {
-    return `${getRadixMarker(Radix.Hexadecimal)}${getAddressString(target, Radix.Hexadecimal)}`;
+export function toHexStringWithRadixMarker(target: bigint, paddedLength: number = 0): string {
```
I love that, great idea!

martin-fleck-at commented on src/common/memory.ts:
```diff
@@ -0,0 +1,74 @@
+export function validateMemoryReference(reference: string): string | undefined {
+    const asNumber = Number(reference);
+    // we allow an address that is not a number, e.g., an expression, but if it is a number it must be >= 0
+    return !isNaN(asNumber) && asNumber < 0 ? 'Value needs to be >= 0' : undefined;
```
As for the other error messages, I simply took them as they were before but I love the more precise and shorter ones that you provided, thank you!

martin-fleck-at commented on src/common/memory.ts:
```diff
@@ -0,0 +1,74 @@
+    memoryReference: string;
+}
+
+export const isMemoryVariable = (variable: unknown): variable is MemoryVariable => !!variable && !!(variable as MemoryVariable).memoryReference;
+
+export interface MemoryVariableNode {
+    variable: MemoryVariable;
+    sessionId: string;
+}
+
+export const isMemoryVariableNode = (node: unknown): node is MemoryVariableNode =>
```
That is interesting! And indeed that changes even a bit of how I think this should be working. The type that we are expecting is coming from the VS Code Variables View ([ref](https://github.com/microsoft/vscode/blob/6c19c003574b02afc604c0c1c471b1441b33a3d3/src/vs/workbench/contrib/debug/browser/variablesView.ts#L56C1-L60C2)):
```
interface IVariablesContext {
	sessionId: string | undefined;
	container: DebugProtocol.Variable | DebugProtocol.Scope | DebugProtocol.EvaluateArguments;
	variable: DebugProtocol.Variable;
}
```
 and for debug variables, the memory reference might not even be set so I wonder if we need a method to get the address for a particular variable somewhere in the debug capabilities. Will check.

martin-fleck-at commented on src/plugin/memory-provider.ts:
```diff
@@ -113,12 +118,22 @@ export class MemoryProvider {
 
-    public async writeMemory(writeMemoryArguments: DebugProtocol.WriteMemoryArguments): Promise<MemoryWriteResult> {
-        return this.assertCapability('supportsWriteMemoryRequest', 'write memory').customRequest('writeMemory', writeMemoryArguments);
+    public async evaluate(args: DebugProtocol.EvaluateArguments): Promise<DebugProtocol.EvaluateResponse['body']> {
+        return sendRequest(this.assertActiveSession('evaluate'), 'evaluate', args);
```
I removed the method for now, thank you!

jreineckearm commented on package.json:
```diff
@@ -92,19 +104,45 @@
           "when": "false"
+        },
+        {
+          "command": "memory-inspector.store-file",
```
We should disable the commands/buttons in the window if no debug session is active.
At the moment, I can click the buttons outside the debug session and get errors as expected.

martin-fleck-at replied:
That is a very good point! 

We can see that the command is already disabled as it does not show up in the command palette anymore (F1 + 'memory' because we can no longer read or write. I moved the enablement condition directly to the command but it seems that VS code still allows us to simply execute the command via code and I don't see any other way to determine whether our command is enabled or not. The documentation even states it like that:

```
(Optional) Condition which must be true to enable the command in the UI (menu and keybindings). Does not prevent executing the command by other means, like the `executeCommand`-api.
```

I added a new commit that now shares the current session state with the memory inspector so we can properly determine the read/write capabilities and disable our buttons accordingly.

martin-fleck-at commented on src/webview/memory-webview-view.tsx:
```diff
@@ -107,6 +112,37 @@ class App extends React.Component<{}, MemoryAppState> {
+            // if we are dealing with numeric addresses (and not expressions) then we can determine the overlap
+            const written: BigIntMemoryRange = {
+                startAddress: BigInt(writtenMemory.memoryReference),
+                endAddress: BigInt(writtenMemory.memoryReference) + BigInt(writtenMemory.count ?? 0)
+            };
+            const shown: BigIntMemoryRange = {
+                startAddress: this.state.memory.address,
+                endAddress: this.state.memory.address + BigInt(this.state.memory.bytes.length)
```
Good question! I think in theory you might be right but in this case we only do an overlap calculation (bigint arithmetic) and we also take the bytes of the current memory and not actually the number of addresses shown. So I believe we are fine even though we have to go with the `endAddress` terminology as it is part of the `BigIntMemoryRange` interface. What do you think? I'll add a comment to avoid confusion.

martin-fleck-at commented:
@colin-grant-work Thank you for your detailed feedback! I tried to address everything in a follow up commit (after a rebase) and added another commit on top based on the feedback from Jens. Besides the type placement, I think I have addressed everything.

jreineckearm commented:
@martin-fleck-at , re-tested functionality.

Did the ultimate test with storing memory from one of our existing debug tools to a hex file. And applying it with the Memory Inspector. And vice versa. Looks all healthy! :-)

The recent change to connect a memory window instance to a debug session ID has the unfortunate side effect of not being able to reuse a memory window anymore. This was possible, would be good to get that back. Use case: I debug something and use the Memory Inspector, I make a code change, then restart debug, and would like to inspect related changes in the previously opened window instance.

Some minor suggestions inline.

One bigger change that may be required and is possibly out of scope for this one is to honor MS-DAP `memory` events. Our debug adapter correctly sends them on memory contents change. But I see that the Memory Inspector debug tracker doesn't make use of them to trigger a refresh if in the loaded memory range of a window instance. At the moment you need to manually trigger a refresh after loading a hex file into memory.

martin-fleck-at commented on src/plugin/memory-storage.ts:
```diff
@@ -0,0 +1,197 @@
+            // if we are already given a URI, let's not bother the user and simply use it
+            return { uri: providedDefault.uri };
+        }
+        const selectedUris = await vscode.window.showOpenDialog({ title: 'Apply Memory', filters: IntelHEX.DialogFilters });
```
Good point, I added that on the last commit!

martin-fleck-at commented on package.json:
```diff
@@ -81,6 +83,18 @@
+      },
+      {
+        "command": "memory-inspector.store-file",
+        "title": "Store Memory as File",
```
Done!

colin-grant-work commented:
Quite a few changes since my last look - I'll read things over later today.

martin-fleck-at commented:
@colin-grant-work Thank you very much! I added them as separate commits so it shouldn't be too bad, I hope! I'll do a quick rebase then.

martin-fleck-at commented:
@colin-grant-work @jreineckearm Everything should be good to review again ;-)

jreineckearm commented:
Thanks, @martin-fleck-at !

From a functionality point of view, things work as outlined in your last comment. This is a really nice new feature!

* Reusing existing and configured Memory Inspector windows for subsequent debug sessions works again.
* The apply/store buttons are now disabled if debug session is not active to prevent users from provoking errors when functionality simply can't work.
* Context menu has been extended by the two commands. Getting crowded there. Need to consider grouping commands by sections/sub-menus when adding new functionality. But I guess we need think about a similar "redesign" for the in-window options menu going forward.
* Apply/store still works as expected.
* Open and save dialogs now start at `${workspaceFolder}`.
* "memory" event support works nicely. I can see the effect now when editing variables and using commands in the Debug Console of our Arm Debugger debug adapter extension.

Minor improvement for later (not this PR)
* We have a context menu entry on the variables view to store data for variables living in memory. But there is no counter-part to quickly load a data set into a variable.

jreineckearm reviewed (approved):
Nice, thanks for incorporating the feedback! LGTM!

martin-fleck-at commented on src/plugin/memory-provider.ts:
```diff
@@ -89,36 +101,63 @@ export class MemoryProvider {
 
-    public async writeMemory(writeMemoryArguments: DebugProtocol.WriteMemoryArguments): Promise<MemoryWriteResult> {
-        return this.assertCapability('supportsWriteMemoryRequest', 'write memory').customRequest('writeMemory', writeMemoryArguments);
+    public async writeMemory(args: DebugProtocol.WriteMemoryArguments & { count?: number }): Promise<WriteMemoryResult> {
```
As it is noted down in the comment below, we want to skip the length calculation if necessary:
```
we accept count as an additional argument so we can skip the memory length calculation
```

The `bytesWritten` is optional for the DAP response and the only other way to get the actual memory length would be to calculate the length on the actual data which may be costly. In cases where we already know how much data we write (e.g., reading from a file) we can skip that calculation by giving the known count already. But we can also skip it, it was really just an optimization.

martin-fleck-at commented on src/common/debug-requests.ts:
```diff
@@ -0,0 +1,72 @@
+    'stopped': DebugProtocol.StoppedEvent
+}
+
+export type DebugRequest<COMMAND, ARGS> = Omit<DebugProtocol.Request, 'command' | 'arguments'> & { command: COMMAND, arguments: ARGS };
+export type DebugResponse<COMMAND, BODY> = Omit<DebugProtocol.Response, 'command' | 'body'> & { command: COMMAND, body: BODY };
```
Thank you, I'll change it.

martin-fleck-at commented on src/common/messaging.ts:
```diff
@@ -17,20 +17,49 @@
+export type ReadMemoryArguments = DebugRequestTypes['readMemory'][0];
+export type ReadMemoryResult = DebugRequestTypes['readMemory'][1];
+
+export type WriteMemoryArguments = DebugRequestTypes['writeMemory'][0] & { count?: number };
```
It is used in when we prepare the "memory written" event when we already know the number of bytes that we are writing. However, maybe this is just too simplistic and I'll remove it for now. Classic case of "premature optimization".

martin-fleck-at commented on src/plugin/memory-provider.ts:
```diff
@@ -14,34 +14,43 @@
+    public readonly onDidWriteMemory = this._onDidWriteMemory.event;
 
-    protected readonly sessions = new Map<string, DebugProtocol.Capabilities | undefined>();
+    private _sessionContext: SessionContext = { canRead: false, canWrite: false };
```
Good catch! I had that in there because my context also supports sending the session id and that id is not part of the `capabilities` response. However, I only had that in for logging and otherwise the session id is not really used so I'll remove it for now. Probably we'll get something similar when we work on #97.

martin-fleck-at replied:
Good catch! I'll remove it.

martin-fleck-at commented on src/plugin/adapter-registry/c-tracker.ts:
```diff
@@ -31,15 +42,14 @@ export class CTracker extends AdapterVariableTracker {
-            const startAddress = BigInt(addressPart[0]);
-            const endAddress = notADigit.test(sizeResponse.result) ? undefined : startAddress + BigInt(sizeResponse.result);
-            this.logger.debug('Resolved', variable.name, { start: addressPart[0], size: sizeResponse.result });
+            const variableAddress = await this.getAddressOfVariable(variable.name, session);
+            if (!variableAddress) { return undefined; }
+            const variableSize = await this.getSizeOfVariable(variable.name, session);
+            if (!variableSize) { return undefined; }
```
That is a good question! It was like that before and I did some quick research: As far as I can tell, there is no "give me details about this variable" request in the DAP, we can only get the children of a variable container. Also I don't think there is a single evaluation that we can give that would yield both. However, I saw that we should be able to query the `variable.memoryReference` and use that as address if it is given. I'll add that.

jreineckearm replied:
@martin-fleck-at , please see also this discussion here regarding serialization and using `memoryReference`: https://github.com/eclipse-cdt-cloud/vscode-memory-inspector/issues/68

Unfortunately it is true that DAP doesn't give us variable information without getting values from the target. Something which would be also useful for extending the variable column search to more than already expanded groups. This is something that may need a DAP change. We currently assume that a DAP supports according `evaluateRequest` expressions. Which should be the case for a C/C++ debugger.

The `memoryReference` could help to skip an actual debug adapter request for at least the variable address though. If it is a number. See Colin's comment here for GDB: https://github.com/eclipse-cdt-cloud/vscode-memory-inspector/issues/68#issuecomment-1952533335

martin-fleck-at replied:
Ah, I see, then I have misunderstood the serialization question. Of course, we can send both requests in parallel and I'll try to skip the address one if we already have a valid hex memory address. @jreineckearm Thank you for the context!

martin-fleck-at commented on src/plugin/external-views.ts:
```diff
@@ -0,0 +1,32 @@
+    }
+}
+
+export function isVariablesContext(context: VariablesView.IVariablesContext | unknown): context is VariablesView.IVariablesContext {
+    const assumed = context ? context as VariablesView.IVariablesContext : undefined;
+    return isDebugVariable(assumed?.variable) && (isDebugVariable(assumed?.container) || isDebugScope(assumed?.container) || isDebugEvaluateArguments(assumed?.container));
+}
```
I think you may have overlooked a parenthesis here, the condition correctly specifies:
```ts
false && (false || false || true) // -> false
```

To be honest, I'm a bit hesitant to ease that check since I believe the VS Code interface gives us a bit more context of  where this data is coming from, i.e., the Variables View. I see that we are only interested in the variable but they are always in the context of a particular user interaction (`Show in Memory Inspector` and `Store Memory to File`) so I'd say that the performance hit for those additional type checks (at most 5 `typeof` executions) is rather low compared to weakening the correctness of the check. How important is that to you?

colin-grant-work replied:
Ah - you're right. I got lost in the function call parentheses and missed the bracketing parentheses.

martin-fleck-at commented:
@colin-grant-work @jreineckearm I pushed an update that should address the latest concerns.

jreineckearm commented on src/plugin/adapter-registry/c-tracker.ts:
```diff
@@ -42,13 +42,14 @@ export class CTracker extends AdapterVariableTracker {
         try {
-            const variableAddress = await this.getAddressOfVariable(variable.name, session);
+            const [variableAddress, variableSize] = await Promise.all([
+                variable.memoryReference && hexAddress.test(variable.memoryReference) ? variable.memoryReference : this.getAddressOfVariable(variable.name, session),
```
Should we consider different integer representations? Not sure if all debug adapters would always return a hex value for an address (although IMO the sensible thing to do).

martin-fleck-at replied:
That is a good point! But I'd rather defer this discussion into a separate issue. The previous behavior also used the `hexAddress`. I did add a commit that unifies that check but I still think it does not necessarily have to be part of this PR, do you agree?

jreineckearm replied:
Fine with me. TBF, I'd expect addresses to be returned as hex values anyway.

colin-grant-work replied:
Since this isn't standardized, I'd probably prefer a check for any (easily parsed) number format so that `16` is valid in addition to `0x10`. That would follow the convention for things like the [`ReadMemoryResponse`](https://microsoft.github.io/debug-adapter-protocol/specification#Requests_ReadMemory) and [`DisassembledInstruction`](https://microsoft.github.io/debug-adapter-protocol/specification#Types_DisassembledInstruction) where at least hex and decimal are both considered acceptable formats.

martin-fleck-at replied:
Fantastic, thank you! I added https://github.com/eclipse-cdt-cloud/vscode-memory-inspector/issues/107 to track this. It shouldn't be too much work but it needs to be discussed first (all formats vs "only" hexadecimal/decimal because they are mentioned in the DAP).

colin-grant-work commented on src/plugin/memory-webview-main.ts:
```diff
@@ -301,4 +312,26 @@ export class MemoryWebview implements vscode.CustomReadonlyEditorProvider {
         this.setMemoryViewSettings(ctx.messageParticipant, { visibleColumns });
     }
+
+    protected async storeMemory(storeArguments: StoreMemoryArguments): Promise<void> {
+        // Even if we disable the command in VS Code through enablement or when condition, programmatic execution is still possible.
+        // However, we want to fail early in case the user tries to execute a disabled command
+        if (!this.memoryProvider.createContext().canRead) {
+            throw new Error('Cannot read memory, no valid debug session.');
+        }
+        return vscode.commands.executeCommand(StoreCommandType, storeArguments);
+    }
+
+    protected async applyMemory(): Promise<MemoryOptions> {
+        // Even if we disable the command in VS Code through enablement or when condition, programmatic execution is still possible.
+        // However, we want to fail early in case the user tries to execute a disabled command
+        if (!this.memoryProvider.createContext().canWrite) {
+            throw new Error('Cannot write memory, no valid debug session.');
+        }
+        return vscode.commands.executeCommand(ApplyCommandType);
+    }
```
For a follow-up: since we're sending a context to the view that includes information about the session that that window is associated with, these requests could handle that association and check the relevant session, rather than defaulting to the active session. Re: #97.

martin-fleck-at replied:
This is a very good point and I believe that should be handled as part of #97 as we definitely want to do this for all requests (read, write, store, apply, get variables).

colin-grant-work reviewed (approved):
I'm happy with the changes. Both of my last comments can be handled as a follow-up, though I'd like it if the address parsing were handled in this PR.

martin-fleck-at commented:
@colin-grant-work Thank you very much! I'll push an update for the decimal address support in a few minutes!

martin-fleck-at commented:
@colin-grant-work  If you agree with the latest commit, please feel free to merge this PR.
